compiler: default to `--gc:orc` for the C target

zerbina opened the pull request:
## Summary

After bootstrapping the compiler itself with `--gc:orc`, this is the
second step towards removing the old runtime.

In order to make to make sure that ORC is stable enough, and that the
standard library works with the new runtime, `--gc:orc` is now implied
for the C target when no memory management strategy is explicitly
selected.

### The New Runtime

Using either `--gc:orc` or `--gc:arc` implies usage of the new runtime,
meaning that the new runtime is now the default. The key differences
compared to the old runtime are:
* usage of `deepcopy` in alive code requires the `--deepcopy` option to
  be enabled
* modifying a string through unsafe facilities (pointers and `cstring`)
  requires preparing the string for mutation first (via
  `prepareMutation`)
* `GC_unref` and `GC_ref` are no longer available for `string` and `seq`
  types
* `repr` version 2 is used

With `--gc:orc` and `--gc:arc`, `shallowCopy` is a no-op.

## Details

* support `prepareMutation` for code running in the VM (it's currently a
  no-op)
* fix a crash with `marshal.loadAny` and empty strings, when using the
  new runtime
* disable tests depending on `nimrtl` -- `nimrtl` doesn't yet work with
  the new runtime
* the IC implementation doesn't support type-bound operators (and thus
  `--gc:arc|orc`) at this time, so change its tests to explicitly use
  `--gc:refc`
* adjust the test suite where necessary

zerbina committed:
vmops: use a no-op wrapper for `prepareMutation`

This allows for code explicitly using `prepareMutation` to run at
compile-time and in the VM in general.

zerbina committed:
compiler: default to `--gc:orc` for the C target

Even though both the JavaScript and VM target don't really use refc,
it still needs to be set as the active memory management strategy
for them.

zerbina committed:
tests: use `--gc:refc` for tests that require it

None of the tests require `refc` *directly* -- rather, they depend on
the old runtime being used.

zerbina committed:
tests: use `--deepcopy:on` where necessary

The option enables generation of the legacy RTTI for all types, the
latter which is necessary in order for the `deepcopy` procedure to
work.

zerbina committed:
tests: explicitly use `--gc:refc`

Until the old runtime is removed, this keeps a tiny amount of test
coverage for `refc`.

zerbina committed:
tests: use `prepareMutation` where required

zerbina committed:
tests: adjust tests depending on `repr` output

`repr` doesn't yield identical results across targets (and it doesn't
need to). Tests that don't strictly require the usage of `repr` are
changed to use other, non-echo based means for checking the invariants.

zerbina committed:
tests: misc cleanup and fixes

zerbina committed:
tests: adjust more tests

* the `GCref` and `GCunref` magics are only available for `ref`s with
  the new runtime
* `repr` works different with the new runtime, so the typedesc tests
  no longe produce the expected results. They're changed to something
  that should (not verified) trigger the failures the associated PRs
  fixed

zerbina committed:
tests: fix the `tatomicptrs.nim` test

It was dependent on undefined behaviour. Namely, that the memory
past the allocated region was accessible and that it stored a
specific value.

zerbina committed:
stdlib: adjust the `mitems` runnable example

Modifying a string through an unsafe view (a `cstring` in this case) requires
one to call `prepareMutation` prior.

zerbina committed:
tests: adjust `tcodegenbug1.nim`

The reaons the code now works as intended is only because the return-
value optimization is not applied as much when using the destructor-
based memory managements schemes.

zerbina committed:
tests: switch the IC category to use `--gc:refc`

zerbina committed:
tests: adjust the `tseq.nim` test

A `shallowCopy` is the same as a normal copy when using the new runtime.
Using a `cursor` best approximates the shallow behaviour.

zerbina committed:
tests: adjust `tmissinginit.nim`

The original issue was specific to `repr`, so the usage of `repr` is
kept for now.

zerbina committed:
tests: fix the `tpackedsets.nim` test

The `typed` template parameter was used in a definition context (likely an
oversight). This lead to the local being treated as a procedure-level
global that required destruction.

A global destructor was thus registered, but since the global is not a
top-level global and destructor call for it was already injected, this
resulted in a double free at run-time.

zerbina committed:
stdlib: small fix for `marshal`

The json deserializing logic made the assumption that strings are
pointer-like, but this is not true with thew runtime. The
implementation-agnostic `setString` is now used.

zerbina committed:
tests: disable the `server` dll test

`nimrtl` is no longer usable with the new runtime, at least until the
relevant parts in `system` are adjusted.

zerbina committed:
tests: fix `tdeepcopy`

The spec was malformed. Testament remembers that an error ocurred, but
overwrites the result code for the test if it's either joined or not in
batch, making these kinds of issues hard to track down.

saem commented on compiler/nim.nim:
```diff
@@ -89,6 +90,18 @@ proc handleCmdLine(cache: IdentCache; conf: ConfigRef): CmdLineHandlingResult =
       conf.errorCounter != 0:
     return
 
+  # if no GC was selected by either the command line or the config files,
```
This needs to be true for suggests as well, I'd put it in `cmdlinehelper` for now.

zerbina replied:
Ah, thanks for catching this. I forgot about that part.

saem commented on tests/stdlib/metaprogramming/mgenast.nim:
```diff
@@ -31,6 +31,7 @@ macro bindme6UseExpose*(): untyped =
   genAst:
     var tst = "sometext"
     var ss = newStringStream("anothertext")
+    prepareMutation(tst)
```
Promise me that if we do something with copy on write it has to be named `prepareMootation`. 🐄

saem commented on tests/stdlib/types/tpackedsets.nim:
```diff
@@ -235,15 +235,15 @@ block legacyMainModuleTests:
       proc bug12366 =
         var
-          x = initPackedSet[A]()
+          a = initPackedSet[A]()
```
No need to change it back, just curious why the rename?

zerbina replied:
It's the symbol re-definition issue again (although in a much more subtle way).

`x` is also the name of a `typed` argument of the template this procedure is defined inside. Below, the template is invoked with non-top-level globals passed to `x`.

This leads to the `var substitutedGlobalSymbolGoesHere` to be treated as a procedure-level global. If those have a destructor (`legacyInit`, `instGenericInit`, etc. do have one), they're registered in the global destructor list, which in this case caused the globals to be destroyed twice (a double free that resulted in a crash).

saem reviewed (commented):
I spotted one thing that needs fixing, selecting the gc if none is set as it impacts what's defined or not in stdlib.

Otherwise this is great, very much looking forward to getting this merged in. Thank you very much for persevering. ♥️🚀

zerbina committed:
nim: move default GC logic to separate procedure

`nimsuggest` (and other tools) needs to perform the same action, so the
logic is moved to a centrally accessible location (the `cmdlinehelper`
module) for now.

saem commented:
/merge
